pass/ProteusPass.cpp

```diff
@@ -868,6 +742,135 @@ class ProteusJitPassImpl {
+        "__jit_register_fatbinary", JitRegisterFatbinaryFnTy);
+
+    return JitRegisterFatbinaryFn;
+  }
+
+  void instrumentRegisterFatBinary(Module &M) {
+    Function *F = nullptr;
+
+    if (!RegisterFatBinaryName)
+      return;
+
+    F = M.getFunction(RegisterFatBinaryName);
+    if (!F)
+      return;
+
+    FunctionCallee JitRegisterFatBinaryFn = getJitRegisterFatBinaryFn(M);
+
+    for (auto *User : F->users()) {
+      CallBase *CB = dyn_cast<CallBase>(User);
+      if (!CB)
+        continue;
+
+      IRBuilder<> Builder(CB->getNextNode());
+      Value *FatbinWrapper = CB->getArgOperand(0);
+
+      std::string GVName = getJitBitcodeUniqueName(M);
+      DEBUG(dbgs() << "Instrument register fatbinary bitcode GV " << GVName
+                   << "\n";);
+      auto *Arg = Builder.CreateGlobalString(GVName);
+
+      Builder.CreateCall(JitRegisterFatBinaryFn, {CB, FatbinWrapper, Arg});
+    }
+  }
+
+  FunctionCallee getJitRegisterFatBinaryEndFn(Module &M) {
+    FunctionType *JitRegisterFatBinaryEndFnTy =
+        FunctionType::get(VoidTy, {PtrTy},
+                          /* isVarArg=*/false);
+    FunctionCallee JitRegisterFatBinaryEndFn = M.getOrInsertFunction(
+        "__jit_register_fatbinary_end", JitRegisterFatBinaryEndFnTy);
+
+    return JitRegisterFatBinaryEndFn;
+  }
+
+  void instrumentRegisterFatBinaryEnd(Module &M) {
+// This is CUDA specific.
+#if !ENABLE_CUDA
+    return;
+#endif
+
+    Function *F = M.getFunction("__cudaRegisterFatBinaryEnd");
+    if (!F)
+      return;
+
+    FunctionCallee JitRegisterFatBinaryEndFn = getJitRegisterFatBinaryEndFn(M);
+
+    for (auto *User : F->users()) {
+      CallBase *CB = dyn_cast<CallBase>(User);
+      if (!CB)
+        continue;
+
+      IRBuilder<> Builder(CB->getNextNode());
+      Value *FatbinWrapper = CB->getArgOperand(0);
+      Builder.CreateCall(JitRegisterFatBinaryEndFn, {FatbinWrapper});
+    }
+  }
+
+  FunctionCallee getJitRegisterLinkedBinaryFn(Module &M) {
+    FunctionType *JitRegisterLinkedBinaryFnTy =
+        FunctionType::get(VoidTy, {PtrTy},
+                          /* isVarArg=*/false);
+    FunctionCallee JitRegisteLinkedBinaryrFn = M.getOrInsertFunction(
+        "__jit_register_linked_binary", JitRegisterLinkedBinaryFnTy);
+
+    return JitRegisteLinkedBinaryrFn;
+  }
+
+  void instrumentRegisterLinkedBinary(Module &M) {
+// This is CUDA specific.
+#if !ENABLE_CUDA
+    return;
+#endif
+
+    // Note: we check for __cuda_fatibn_wrapper to avoid emitting for the
+    // link.stub. It's not strictly necessary since this module will not have a
+    // device bitcode to pull and we skip at runtime.
+    if (!M.getGlobalVariable("__cuda_fatbin_wrapper", /*AllowInternal=*/true)) {
+      DEBUG(dbgs() << "Skip " << M.getSourceFileName() << "\n";)
+      return;
+    }
+
+    FunctionCallee JitRegisterLinkedBinaryFn = getJitRegisterLinkedBinaryFn(M);
+
+    for (auto &F : M.getFunctionList()) {
```

Comment:
Instead of iterating all functions, get the handle `__cuda_module_ctor`. Inside that function __cudaRegisterLinkedBinary is being called. The constructor is small, so in principle it should be faster.

Reply:
This is a good suggestion. I'll defer for iteration 2 and open an issue.